storage/remote/otlptranslator/prometheusremotewrite/histograms.go

```diff
@@ -226,3 +237,93 @@ func convertBucketsLayout(buckets pmetric.ExponentialHistogramDataPointBuckets,
 
 	return spans, deltas
 }
+
+func (c *PrometheusConverter) addCustomBucketsHistogramDataPoints(ctx context.Context, dataPoints pmetric.HistogramDataPointSlice,
+	resource pcommon.Resource, settings Settings, promName string) (annotations.Annotations, error) {
+	var annots annotations.Annotations
+
+	for x := 0; x < dataPoints.Len(); x++ {
+		if err := c.everyN.checkContext(ctx); err != nil {
+			return annots, err
+		}
+
+		pt := dataPoints.At(x)
+
+		histogram, ws, err := explicitHistogramToCustomBucketsHistogram(pt)
+		annots.Merge(ws)
+		if err != nil {
+			return annots, err
+		}
+
+		lbls := createAttributes(
+			resource,
+			pt.Attributes(),
+			settings,
+			nil,
+			true,
+			model.MetricNameLabel,
+			promName,
+		)
+
+		ts, _ := c.getOrCreateTimeSeries(lbls)
+		ts.Histograms = append(ts.Histograms, histogram)
+
+		exemplars, err := getPromExemplars[pmetric.HistogramDataPoint](ctx, &c.everyN, pt)
+		if err != nil {
+			return annots, err
+		}
+		ts.Exemplars = append(ts.Exemplars, exemplars...)
+	}
+
+	return annots, nil
+}
+
+func explicitHistogramToCustomBucketsHistogram(p pmetric.HistogramDataPoint) (prompb.Histogram, annotations.Annotations, error) {
+	var annots annotations.Annotations
+
+	buckets := p.BucketCounts().AsRaw()
+	offset := getBucketOffset(buckets)
+	bucketCounts := buckets[offset:]
+	positiveSpans, positiveDeltas := convertBucketsLayout(bucketCounts, int32(offset), 0, false)
+
+	h := prompb.Histogram{
+		// The counter reset detection must be compatible with Prometheus to
+		// safely set ResetHint to NO. This is not ensured currently.
+		// Sending a sample that triggers counter reset but with ResetHint==NO
+		// would lead to Prometheus panic as it does not double check the hint.
+		// Thus we're explicitly saying UNKNOWN here, which is always safe.
+		// TODO: using created time stamp should be accurate, but we
+		// need to know here if it was used for the detection.
+		// Ref: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/28663#issuecomment-1810577303
+		// Counter reset detection in Prometheus: https://github.com/prometheus/prometheus/blob/f997c72f294c0f18ca13fa06d51889af04135195/tsdb/chunkenc/histogram.go#L232
+		ResetHint: prompb.Histogram_UNKNOWN,
+		Schema:    -53,
+
+		PositiveSpans:  positiveSpans,
+		PositiveDeltas: positiveDeltas,
+		CustomValues:   p.ExplicitBounds().AsRaw(),
```

Comment:
We don't encode the +Inf upper bound in NHCB, so we'll need an extra step to remove it if Otel had it. See https://prometheus.io/docs/specs/native_histograms/#custom-values

Reply:
I believe that OTel explicit histograms use an implicit +Inf bucket. Added a comment to the code.